ast/spec/earthfile.go

```diff
@@ -0,0 +1,77 @@
+package spec
+
+// Earthfile is the AST representation of an Earthfile.
+type Earthfile struct {
+	BaseRecipe     Block           `json:"baseRecipe"`
+	Targets        []Target        `json:"targets"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// Target is the AST representation of an Earthfile target.
+type Target struct {
+	Name           string          `json:"name"`
+	Recipe         Block           `json:"recipe"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// Block is the AST representation of a block of statements.
+type Block []Statement
+
+// Statement is the AST representation of an Earthfile statement. Only one field may be
+// filled at one time.
+type Statement struct {
+	Command        *Command        `json:"command,omitempty"`
+	With           *WithStatement  `json:"with,omitempty"`
+	If             *IfStatement    `json:"if,omitempty"`
+	ForIn          *ForInStatement `json:"forIn,omitempty"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// Command is the AST representation of an Earthfile command.
+type Command struct {
+	Name           string          `json:"name"`
+	Args           []string        `json:"args"`
+	ExecMode       bool            `json:"execMode,omitempty"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// WithStatement is the AST representation of a with statement.
+type WithStatement struct {
+	Command        Command         `json:"command"`
+	Body           Block           `json:"body"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// IfStatement is the AST representation of an if statement.
+type IfStatement struct {
+	Expression     []string        `json:"expression"`
+	IfBody         Block           `json:"ifBody"`
+	ElseIf         []ElseIf        `json:"elseIf"`
+	HasElse        bool            `json:"hasElse"`
+	ElseBody       *Block          `json:"elseBody,omitempty"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// ElseIf is the AST representation of an else if clause.
```

Comment:
Do we need to represent an ElseIF in the AST? I think it's good to support it in the parser, but by the time the interpreter receives it would it simplify the interpreter code if it was simple an IFBody and ElseBody, then each ElseBody would container another IfStatement nested under it.

Reply:
I was debating this in my head. It's not absolutely necessary. I figured it looks more like an AST if there is less interpreting in the listener. But perhaps that is not strictly the goal here - but rather to have some lower-level representation of the operations.

Reply:
I was just thinking out loud here. If we keep it, does it make it easier to display clear error messages?

Reply:
Maybe... it definitely maps more closely to the original code. But it's not a big difference.

It may be more significant that for us it's easier to read the AST for debug purposes, as there's less JSON nesting.

Reply:
I ended up removing it - we can add it back in the future if we change our minds.